### ISSUE TITLE
Caching issue

### ISSUE THREAD

**Issue by notEvil** [sgsokol/Deriv#5]
```
f = quote(
(1) * (({
    .e3 <- exp(-(theta0 + theta2 * x))
    .e4 <- exp(-(theta0 + theta3 * x))
    .e5 <- .e4 - .e3
    .e6 <- gamma1^tau
    2 * ((1 - beta2/.e5) * exp(-(.e6 * .e4))) + 2 * (beta2 * 
        exp(-(.e6 * .e3))/.e5) - 1
}) * ({
    .e3 <- exp(-(theta0 + theta1 * x))
    .e4 <- exp(-(theta0 + theta3 * x))
    .e5 <- .e4 - .e3
    .e6 <- gamma2^tau
    2 * ((1 - beta1/.e5) * exp(-(.e6 * .e4))) + 2 * (beta1 * 
        exp(-(.e6 * .e3))/.e5) - 1
}))
)

Deriv::Deriv(f, 'beta1')
```

the issue is in the second statement of the result

```
.e6.e4 <- exp(-(.e3 * .e6))
```

**Comment by sgsokol** [comment]
Could you elaborate a little bit where the expression f comes from? It is strangely formated and not simplified. For the moment, I am not sure that this kind of expression should be taken into account.

**Comment by notEvil** [comment]
it was built using `substitute` and a valid R expression as far as I know. Why would this expression be any different or invalid? Of course it contains some unnecessary brackets, but it is a sub product and nothing the user of my package wants to specify.

**Comment by sgsokol** [comment]
what is annoying for me it's the presеnce of {}-expressions inside of ariythmetic operators (or any other call like exp() etc.) Even if such operator topology is perfectly legal from R point of view, it was not considered during the design of Cache() which is not a general purpose code but is just supposed to treat expressions coming from Deriv_()

**Comment by sgsokol** [comment]
With the last commit in devel (fdf959) you can try to see if it helps:

```
Deriv(deCache(f), "beta1")
{
 .e2 <- exp(-(theta0 + theta3 * x))
 .e4 <- exp(-(theta0 + theta2 * x))
 .e6 <- exp(-(theta0 + theta1 * x))
 .e7 <- .e2 - .e4
 .e8 <- gamma1^tau
 .e9 <- gamma2^tau
 (2 * ((1 - beta2/.e7) * exp(-(.e8 * .e2))) + 2 * (beta2 * 
    exp(-(.e8 * .e4))/.e7) - 1) * (2 * exp(-(.e9 * .e6)) - 
    2 * exp(-(.e9 * .e2)))/(.e2 - .e6)
}
```

**Comment by notEvil** [comment]
sorry for the late response. deCache seemed to work and it might prove useful when dealing with expressions in general. for now I successfully avoided the scenario of {} inside expression.

anyways, I found new issues while using the latest version (3.5.5). see for example

```
a = quote({
    a <- x^2
    b <- x^2
    0
})

Deriv::Deriv(a, 'beta3')

a = quote({
    .e1 <- beta1 + x * (beta2 + beta3 * x)
    .e2 <- x^2
    .e4 <- x * (beta4 + .e2 * (beta5 + beta6 * x))
    .e5 <- 1 + alpha
    .e6 <- pnorm(y1, .e1, 1)
    .e7 <- pnorm(y2, .e4, 1)
    .e8 <- 1/alpha
    .e9 <- .e8 + 2
    .e12 <- 1/.e6^alpha + 1/.e7^alpha - 1
    .e13 <- .e7^.e5
    .e14 <- .e12^.e9
    .e15 <- dnorm(y1, .e1, 1)
    alpha * .e2 * (1 + .e8) * ((.e5/(.e14 * .e6^(2 + alpha) * 
        .e13) - alpha * .e9/(.e12^(.e8 + 3) * .e6^(2 * .e5) * 
        .e13)) * .e15 + (y1 - .e1)/(.e14 * .e6^.e5 * .e13)) * 
        .e15 * dnorm(y2, .e4, 1)
})

Deriv::Deriv(a, 'beta3')

a = quote({
    .e1 <- beta3 * x^2
    .e2 <- x^2
    .e6 <- .e1*2
    .e12 <- .e6^(-alpha)
    #.e2 * .e6^(2 + alpha)
    x^2 * ((beta3 * x^2)*2)^(2 + alpha)
})

Deriv::Deriv(a, 'beta3')
```

the first one obviously is odd but something one can avoid. the second one (a real life example) uses .e2 but never defines it. and the last one is something I came up with when I experimentally tried to figure out what Deriv does.

Last but not least, you probably have seen that `docopulae` now suggests your package. It's the one I talked about a few months ago. It really benefits from speed and accuracy gains when it relies on Deriv instead of numDeriv.

**Comment by sgsokol** [comment]
Le 16/09/2015 21:44, notEvil a écrit :

> sorry for the late response. deCache seemed to work and it might prove useful when dealing with expressions in general. for now I successfully avoided the 
> scenario of {} inside expression.
> 
> anyways, I found new issues while using the latest version (3.5.5). see for example
> 
> |a = quote({ a <- x^2 b <- x^2 0 }) Deriv::Deriv(a, 'beta3')|
> |Nice shot. When I've fixed this one, I could not see any more
> the bugs in the next two. Could you check that it's OK on your side
> with the latest version in devel branch please?

|
|

> pnorm(y2, .e4, 1) .e8 <- 1/alpha .e9 <- .e8 + 2 .e12 <- 1/.e6^alpha + 1/.e7^alpha - 1 .e13 <- .e7^.e5 .e14 <- .e12^.e9 .e15 <- dnorm(y1, .e1, 1) alpha \* .e2 \* 
> (1 + .e8) \* ((.e5/(.e14 \* .e6^(2 + alpha) \* .e13) - alpha \* .e9/(.e12^(.e8 + 3) \* .e6^(2 \* .e5) \* .e13)) \* .e15 + (y1 - .e1)/(.e14 \* .e6^.e5 \* .e13)) \* .e15 \* 
> dnorm(y2, .e4, 1) }) Deriv::Deriv(a, 'beta3') a = quote({ .e1 <- beta3 \* x^2 .e2 <- x^2 .e6 <- .e1*2 .e12 <- .e6^(-alpha) #.e2 \* .e6^(2 + alpha) x^2 \* ((beta3 
> - x^2)*2)^(2 + alpha) }) Deriv::Deriv(a, 'beta3') |
> 
> the first one obviously is odd but something one can avoid. the second one (a real life example) uses .e2 but never defines it. and the last one is something 
> I came up with when I experimentally tried to figure out what Deriv does.
> 
> Last but not least, you probably have seen that |docopulae| now suggests your package. It's the one I talked about a few months ago.
> 
> Congratulations for this release.
> 
> It really benefits from speed and accuracy gains when it relies on Deriv instead of numDeriv.
> 
> Nice to hear it. Thank you.
> I've seen in the docopulae's doc that the use of Deriv makes the user to map list members
> to vars, e.g. theta[["beta3"]] to beta3. Do you have a small "real world" code sample
> with list members to see if I can differentiate them "as is"?

**Comment by notEvil** [comment]
Thank you! devel version works for at least one example :)
Tomorrow I will check a few more.

The real world code sample you asked for:

```
tt = quote(
log(((1 + ((pnorm(y[[1]], theta[["beta1"]] + theta[["beta2"]] * 
    theta[["x"]] + theta[["beta3"]] * theta[["x"]]^2, 1))^(-theta[["alpha"]]) - 
    1 + (pnorm(y[[2]], theta[["beta4"]] * theta[["x"]] + theta[["beta5"]] * 
    theta[["x"]]^3 + theta[["beta6"]] * theta[["x"]]^4, 1))^(-theta[["alpha"]]) - 
    1))^(((-1/theta[["alpha"]]) - 1) - 1) * (((-1/theta[["alpha"]]) - 
    1) * ((pnorm(y[[2]], theta[["beta4"]] * theta[["x"]] + theta[["beta5"]] * 
    theta[["x"]]^3 + theta[["beta6"]] * theta[["x"]]^4, 1))^((-theta[["alpha"]]) - 
    1) * (-theta[["alpha"]]))) * ((-1/theta[["alpha"]]) * ((pnorm(y[[1]], 
    theta[["beta1"]] + theta[["beta2"]] * theta[["x"]] + theta[["beta3"]] * 
        theta[["x"]]^2, 1))^((-theta[["alpha"]]) - 1) * (-theta[["alpha"]])))) * 
    ((dnorm(y[[1]], theta[["beta1"]] + theta[["beta2"]] * theta[["x"]] + 
        theta[["beta3"]] * theta[["x"]]^2, 1)) * (dnorm(y[[2]], 
        theta[["beta4"]] * theta[["x"]] + theta[["beta5"]] * 
            theta[["x"]]^3 + theta[["beta6"]] * theta[["x"]]^4, 
        1))))
)

Deriv::Deriv(tt, 'theta[["beta1"]]')
```

don't know if implementing the required behavior makes any sense though.

**Comment by notEvil** [comment]
Didn't experience any further issues. Well done :)
I will remove the hint in docopulae's README as soon as the fix is on CRAN.

**Comment by sgsokol** [comment]
bug fixed in v3.5.6 which is on cran now

**Comment by sgsokol** [comment]
Hi,

I have added a possibility to derivate by list/vector components. The syntax to indicate that a given name must be taken in some list (or vector) is based on names in the argument x of Deriv(), e.g. x=c(theta="m")
Here the name "theta" is list name, and the character string "m" is the name of the corresponding component.
If you have all you paramaters in a list theta and you want to diffirentiate by all of them some expression  then you can simply do:
 theta=list(m=0.1, sd=2)
 x=names(theta)
 names(x)=rep_len("theta", length(theta))
 Deriv(~pnorm(q, theta$m, theta$sd), x)

If you want to give it a ride, currently, it is in devel branch (future v3.6.0))

Hoping it helps,
Serguei.

**Comment by notEvil** [comment]
Hello,

just recently I checked on Deriv to see if you pushed this change to release branch and CRAN. I was quite surprised that you haven't been idle and improved it beyond expectation. So I decided to change docopulae in favor of Deriv. That is, it uses Deriv to simplify and cache expressions as well as to compute derivatives involving list components. It's now on CRAN. So thanks for your effort :)

Best, Andreas
